### PR TITLE
Fix NPE using mock for unknown database objects.

### PR DESCRIPTION
**Pull request by charlesmchen-signal**
PTAL @michaelkirk

### CODE REVIEW

File: SignalServiceKit/src/Storage/OWSStorage.m [hunk 1]
```diff
@@ -457,7 +482,8 @@ + (YapDatabaseDeserializer)logOnFailureDeserializer
 
     return ^id(NSString __unused *collection, NSString __unused *key, NSData *data) {
         if (!data || data.length <= 0) {
-            return nil;
+            OWSProdLogAndFail(@"%@ can't deserialize null object: %@", self.logTag, collection);
+            return [OWSUnknownDBObject new];
```
Comment (charlesmchen-signal) [inline SignalServiceKit/src/Storage/OWSStorage.m]: Just being defensive and trying to eliminate other possible causes of NPE.  But I don't think this is the issue after all.

File: SignalServiceKit/src/Storage/OWSStorage.m [hunk 2]
```diff
@@ -205,17 +205,42 @@ @interface OWSUnknownDBObject : NSObject <NSCoding>
...
+}
+
+- (void)touchWithTransaction:(YapDatabaseReadWriteTransaction *)transaction
+{
+    OWSProdLogAndFail(@"%@ Tried to touch unknown object", self.logTag);
+
+    // No-op.
+}
+
+- (void)removeWithTransaction:(YapDatabaseReadWriteTransaction *)transaction
```
Comment (charlesmchen-signal) [inline SignalServiceKit/src/Storage/OWSStorage.m]: This should make our mock slightly safer.

File: SignalServiceKit/src/Contacts/TSThread.m [hunk 3]
```diff
@@ -97,12 +97,14 @@ - (void)removeAllThreadInteractionsWithTransaction:(YapDatabaseReadWriteTransact
...
     OWSAssert(interactionsByThread);
-    [interactionsByThread
-        enumerateKeysInGroup:self.uniqueId
-                  usingBlock:^(
-                      NSString *_Nonnull collection, NSString *_Nonnull key, NSUInteger index, BOOL *_Nonnull stop) {
-                      [interactionIds addObject:key];
-                  }];
+    [interactionsByThread enumerateKeysInGroup:self.uniqueId
+                                    usingBlock:^(NSString *collection, NSString *key, NSUInteger index, BOOL *stop) {
+                                        if (key.length < 1) {
```
Comment (charlesmchen-signal) [inline SignalServiceKit/src/Contacts/TSThread.m]: I noticed that its the _key_ not the _deserialized object_ that is invalid (probably nil).  Which looks a lot like database corruption.

Comment (michaelkirk-signal) [reply SignalServiceKit/src/Contacts/TSThread.m]: ah, good catch. Specifically - seems like our view is out of sync with the backing collection. 

Should we trigger a rebuild?

Comment (charlesmchen-signal) [reply SignalServiceKit/src/Contacts/TSThread.m]: > Should we trigger a rebuild?

Done.

**Pull request merged**